### PR TITLE
Create CustomJob and Datasets operators for Vertex AI service

### PR DESCRIPTION
**Pull request by MaksYermak**
Create operators for working with Custom Job and Datasets for Vertex AI service. Includes operators, hooks, example dags, tests and docs.

Co-authored-by: Wojciech Januszek januszek@google.com
Co-authored-by: Lukasz Wyszomirski wyszomirski@google.com
Co-authored-by: Maksim Yermakou maksimy@google.com

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (josh-fell) [inline airflow/providers/google/cloud/example_dags/example_vertex_ai.py]: Should this be "GCP_VERTEX_AI_BUCKET" instead of "GCP_BUCKET_NAME"?

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (josh-fell) [inline airflow/providers/google/cloud/example_dags/example_vertex_ai.py]: Applicable to both DAGs in this file:

- Example DAGs should use a static `start_date` value instead of a dynamic one as best practice. The value doesn't matter.

- Also, we've been adding `catchup=False` to initially ward off new-user headaches when these example DAGs are copied and `start_date` is modified to fit needs.

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (josh-fell) [inline airflow/providers/google/cloud/example_dags/example_vertex_ai.py:123]: Right now all of these task will execute in parallel and have no dependencies on each other. If there is a use case in which some of these tasks _could_ flow together in a pipeline it would be nice to demonstrate here.

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/example_dags/example_vertex_ai.py]: @josh-fell yes you right, I have changed it

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/example_dags/example_vertex_ai.py]: @josh-fell I have changed it

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/example_dags/example_vertex_ai.py:123]: @josh-fell right now we don't have any dependencies between these task, but in the future iterations maybe we will add some dependencies.

File: scripts/ci/pre_commit/pre_commit_check_provider_yaml_files.py
Comment (MaksYermak) [inline scripts/ci/pre_commit/pre_commit_check_provider_yaml_files.py:157]: I have done this changes, because for Vertex AI we use package for code rather than one module and for packages our static check doesn't work.

**Comment by potiuk** [comment]
Tests are failing - please fix and rebase.

**Comment by MaksYermak** [comment]
@potiuk could we merge this PR ?

**Comment by MaksYermak** [comment]
@josh-fell @mik-laj @turbaszek @vikramkoka @ashb @potiuk  hi guys what about this PR, could we merge it?

File: airflow/providers/google/cloud/operators/vertex_ai/custom_job.py
Comment (josh-fell) [inline airflow/providers/google/cloud/operators/vertex_ai/custom_job.py]: FYI - As of Airflow 2.2.3, there is a specific `Context` object available to use for typing but to make sure the operator doesn't have an implicit Airflow-version dependency you'll need to use `TYPE_CHECKING`.

Something like:
```python
if TYPE_CHECKING:
    from airflow.utils.context import Context
...

def execute(self, context: "Context"): ...
```
This can be done for `ListCustomTrainingJobOperator` and the Dataset operators as well.

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (josh-fell) [inline airflow/providers/google/cloud/example_dags/example_vertex_ai.py:188]: In the operator doc, this operator and a couple others will "...wait for the operation to complete" but the example shows this operator running asynchronously. Should this operator and the others mentioned in the doc _always_ execute synchronously (it looks like `sync` is forced to `True` when calling the underlying hook method in the operator) or is configurability allowed for running the job in an async manner?

**Review by josh-fell — COMMENTED** [review]
It would be nice if the Custom Job operators had docstrings for params/types so they are exposed in the Airflow documentation.

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/example_dags/example_vertex_ai.py:188]: This `sync` relates to aiplatfrom job not to our operator. Our hook runs this job with specific parameters and than this job create CustomJob inside Google Cloud after that our operator waits for CustomJob finished. Our operator works as usual.

File: airflow/providers/google/cloud/operators/vertex_ai/custom_job.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/operators/vertex_ai/custom_job.py]: I have changed Dict to Context in the last commit

**Comment by MaksYermak** [comment]
@potiuk @josh-fell hi guys could you look on my PR one more time and approve/merge if all looks good?

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (josh-fell) [reply airflow/providers/google/cloud/example_dags/example_vertex_ai.py:188]: I think my confusion is in the `sync` parameter description of "...execute this method synchronously". The "method" I guess is some underyling API method from what you're saying. I assumed the description was for the operator but if it's related to the job running on AI Platform, maybe it's worth updating slightly? WDYT?

Something like "Whether to execute the AI Platform job synchronously...."?

**Comment by josh-fell** [comment]
> @potiuk @josh-fell hi guys could you look on my PR one more time and approve/merge if all looks good?

Just a small comment on a possible clarification but LGTM.

File: airflow/providers/google/cloud/example_dags/example_vertex_ai.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/example_dags/example_vertex_ai.py:188]: I think it is good idea. I will update docstring in next PR

**Comment by josh-fell** [comment]
@ashb Should the `:type:` directives in the docstrings here be removed preemptively because of #20951?

**Comment by ashb** [comment]
Yes please!

**Comment by MaksYermak** [comment]
> @ashb Should the `:type:` directives in the docstrings here be removed preemptively because of #20951?

I have deleted :type in last commit

**Comment by ashb** [comment]
@MaksYermak You might have to add some local spelling dictionaries to your hook file too -- check out https://github.com/apache/airflow/pull/20951/commits/1149e63e45d9e5047935c11287a4d0b23e87d2bd

Sorry about that.

**Comment by MaksYermak** [comment]
> @MaksYermak You might have to add some local spelling dictionaries to your hook file too -- check out [1149e63](https://github.com/apache/airflow/commit/1149e63e45d9e5047935c11287a4d0b23e87d2bd)
> 
> Sorry about that.

@ashb I have run this command for check `./breeze build-docs -- --spellcheck-only --package-filter apache-airflow-providers-google` and I haven´t seen any spell errors in this code.

**Comment by MaksYermak** [comment]
@ashb @josh-fell @potiuk Hi all, could we merge this PR?

**Review by potiuk — APPROVED** [review]
Big. But looks like really all things are there. Was it - in big parts auto-generated? It certainly looks like (and we used to have airflow-munchkin for that :)), so I wonder if this is some kind of successor.

**Comment by jedcunningham** [comment]
This had a bunch of failing mypy issues once it landed in main. #21203 reverted it.

@MaksYermak, can you rebase your changes on current main, fix the mypy issues, and open a new PR? Thanks!

**Comment by potiuk** [comment]
Yeah. So the up-to-date checker did not prevent it because it was really last run 8 hours ago :( . Up-to-date was pretty useless actually :(

**Comment by MaksYermak** [comment]
@potiuk @jedcunningham I have created a new PR https://github.com/apache/airflow/pull/21253

**Pull request merged**
